cEP-0018.md

```diff
@@ -0,0 +1,245 @@
+
+| Metadata |                                               |
+| -------- | --------------------------------------------- |
+| cEP      | 18                                            |
+| Version  | 1.0                                           |
+| Title    | Integration of ANTLR into coala core          |
+| Authors  | Viresh Gupta <mailto:viresh16118@iiitd.ac.in> |
+| Status   | Proposed                                      |
+| Type     | Feature                                       |
+
+## Abstract
+
+This document describes how an API based on ANTLR will be constructed and
+maintained.
+
+## Introduction
+
+ANTLR provides parsers for various language grammars and thus we can provide
+an interface from within coala and make it available to bear writers so that
+they can write advanced linting bears. This will aim at supporting the more
+flexible visitor based method of AST traversal (as opposed to listener based
+mechanisms).
+
+The proposal is to introduce a parallel concept to the coala-bears library,
+which will be called the coala-ast hereon.
+
+## Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. There will be a separate repository named as coala-antlr which will
+   be installable via `pip install coala-antlr`.
+2. A new package would be introduced in the coala-antlr repository,
+   which would be the maintain all the dependencies for antlr-ast based
+   bears. This would be the `coantlib` package
+3. The `coantlib` package will provide endpoints relevant to the
+   visitor model of traversing the ast generated via ANTLR.
+4. Another package named `coantparsers` will be created inside
+   coala-antlr repository. This package will hold parsers and lexers generated
+   for various languages generated (with python target) beforehand for the
+   supported set of grammars.
+5. `coantlib` will have an `ASTLoader` class that will be responsible
+   for loading the AST of a given file depending on the user input, using file
+   extension as a fallback.
+   For e.g, a .c file will be loaded using the parser from
+   `coantlib` for c language.
+6. Another `ASTWalker` class will be responsible for providing a single method
+   to resolve the language supplied by the bear and create a walker instance
+   of the appropriate type.
+7. Also coantlib.walkers will have several walker classes derived from
+   `ASTWalker` class that will be language specific and grammar dependant.
+   For e.g a `Py3ASTWalker`.
+8. The bears dependant on the `coantlib` will reside in the same repository as
+   the coantlib. All of them will be in a different package, the coant-bears
+   package, all of them installable together via a single option in the
+   setup.py along with coantlib.
+9. The bears will be able to use a high level api by defining the LANGUAGE
+   instance variable, which will return to them a walker with high level api's
+   in case the language is supported. (for e.g if LANGUAGE='Python', they will
```

Comment:
LANGUAGE='Python' needs backtick, so does LANGUAGE above, and so does any class name anywhere in the prose, such as BasePyASTWalker on the next line